### PR TITLE
Px4 toc enhancement

### PR DESCRIPTION
**Pull request by pussuw**
Move ToC out from the firmware itself, turn it into more like a firmware header

### CODE REVIEW

File: platforms/nuttx/src/bootloader/common/image_toc.c [hunk 1]
```diff
@@ -52,7 +52,7 @@
 
 bool find_toc(const image_toc_entry_t **toc_entries, uint8_t *len)
 {
-	const uintptr_t toc_start_u32 = APP_LOAD_ADDRESS + BOOT_DELAY_ADDRESS + 8;
+	const uintptr_t toc_start_u32 = FLASH_START_ADDRESS + BOOT_DELAY_ADDRESS + 8;
 	const image_toc_start_t *toc_start = (const image_toc_start_t *)toc_start_u32;
 	const image_toc_entry_t *entry = (const image_toc_entry_t *)(toc_start_u32 + sizeof(image_toc_start_t));
```
Comment (jlaitine) [inline platforms/nuttx/src/bootloader/common/image_toc.c:58]: No really related to this PR, bug while on it, could you also add the mitigation to jira DP-2949? Namely this:

- Apply a check to ensure that the TOC boundaries lie within the boundaries specified by the first entry.

No need to do the other "additional hardenings" or such mentioned in the jira.

For more info, see: https://ssrc.atlassian.net/browse/DP-2949

Just that if going to change/upstream the change, this saves some extra burden in the future.

**Comment by jlaitine** [comment]
Looks like a good approach to me

File: boards/ssrc/icicle/nuttx-config/scripts/script.ld [hunk 2]
```diff
@@ -27,12 +27,15 @@ MEMORY
...
 ENTRY(_stext)
-
 EXTERN(__start)
-EXTERN(_main_toc)
 
 SECTIONS
 {
+    .toc : {
+        /* Make room for the ToC */
+        . = 0x80;
```
Comment (pussuw) [inline boards/ssrc/icicle/nuttx-config/scripts/script.ld:36]: 0x40 = 64 bytes, need to shrink this more still

File: boards/ssrc/icicle/nuttx-config/scripts/toc.ld [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+EXTERN(_main_toc)
+
+SECTIONS
+{
+    .firmware : {
+        _fw_start = ABSOLUTE(.);
+
+        /* The ToC */
+        *(.main_toc)
+        . = 0x80;
```
Comment (pussuw) [inline boards/ssrc/icicle/nuttx-config/scripts/toc.ld:39]: Same issue

File: platforms/nuttx/src/bootloader/common/crypto.c [hunk 4]
```diff
@@ -49,10 +51,19 @@ bool verify_app(uint16_t idx, const image_toc_entry_t *toc_entries)
 	uint8_t sig_key = toc_entries[idx].signature_key;
 	crypto_session_handle_t handle = crypto_open(BOOTLOADER_SIGNING_ALGORITHM);
 	app_signature_ptr = (volatile uint8_t *)toc_entries[sig_idx].start;
-	len = (size_t)toc_entries[idx].end - (size_t)toc_entries[idx].start;
+
+	/* For index 0 the ToC is part of the signature */
+	if (idx == 0) {
```
Comment (pussuw) [inline platforms/nuttx/src/bootloader/common/crypto.c:56]: This should fix https://ssrc.atlassian.net/browse/DP-2949

File: boards/ssrc/icicle/nuttx-config/scripts/script.ld [hunk 5]
```diff
@@ -20,16 +20,15 @@
 
 MEMORY
 {
-  progmem (rx) : ORIGIN = 0xAFB00000, LENGTH = 3072K  /* w/ cache */
+  toc     (rx) : ORIGIN = 0xAFB00000. LENGTH = 64 /* w/ cache */
```
Comment (pussuw) [inline boards/ssrc/icicle/nuttx-config/scripts/script.ld:23]: Nope, this size is simply not enough. #define MAX_TOC_ENTRIES 32, one entry is 32 octets. So 1K + a few octets extra is needed.

Back to the drawing board I guess..

**Pull request closed without merging**